torchmultimodal/architectures/omnivore.py

```diff
@@ -0,0 +1,39 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from torch import nn
+
+
+class OmnivoreArchitecture(nn.Module):
+    """Omnivore is a model that accept multiple vision modality.
+
+    Omnivore (https://arxiv.org/abs/2201.08377) is a single model that able to do classification
+    on images, videos, and single-view 3D data using the same shared parameters of the encoder.
+
+    Args:   encoder (nn.Module): Instantiated encoder.
+                See SwinTransformer3dEncoder class.
+            heads (Optinal[nn.ModuleDict]): Dictionary of multiple heads for each dataset type
+
+    Inputs: x (Tensor): 5 Dimensional batched video tensor with format of B C D H W
+                where B is batch, C is channel, D is time, H is height, and W is width.
+            input_type (str): The dataset type of the input, this will used to choose
+                the correct head.
+    """
+
+    def __init__(self, encoder: nn.Module, heads: nn.ModuleDict):
+        super().__init__()
+        self.encoder = encoder
+        self.heads = heads
+        self.input_types = set(heads.keys())
+
+    def forward(self, x: torch.Tensor, input_type: str):
```

Comment:
The idiom of passing the `input_type` as string and then choosing to which head to forward the data is not very common. Why can't we instantiate Omnivore modules with the same encoder but different heads and push to the training loop which module you use depending on the input? This will move the complexity from the nn.Module to the loop and it will be beneficial for use-cases where the loss is different for each head. 

Speaking offline with @YosuaMichael there are some concerns on whether the whole approach will work well on a distributed setup.

Reply:
Hi @datumbox, currently I plan to try on doing the training first and see if there are any particular problem with the architectures (whether the current or multiple model with shared encoder). In particular, I need to check the behaviour of having 2 models `A` and `B` that shared the same `encoder` wrapped individually into `DDP(A)` and `DDP(B)`.

Reply:
@ankitade yes, the original checkpoint include all the heads

Reply:
@ankitade @YosuaMichael It's worth keeping in mind that this pattern wont support FX.

FX does tracing which means the flow of the execution of the model should not depend on the input. By adding the string FX won't know what to execute. Even though you might not be interested right now to make it FX traceable, on the future you might want to adopt FX quantization or other FX-based utils.

This is the reason my advice is to split this module in submodules depending on the head. It's a safer idiom that would be forward compatible with future core expansions. It might require some massaging on the original weights to fix but I think that's straightforward to do and worth it.

Up to you. :)